### PR TITLE
Remove type declaration for Webhooks and Supression classes

### PR DESCRIPTION
**Pull request by ixarlie**
Resolves https://github.com/mailgun/mailgun-php/issues/783

### CODE REVIEW

File: src/Api/Suppression.php [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 namespace Mailgun\Api;
 
+use Http\Client\Common\PluginClient;
```
Comment (Nyholm) [inline src/Api/Suppression.php:14]: Lets use `Http\Client\HttpClient` instead.

**Review by DavidGarciaCat — COMMENTED** [review]
There is also a BC break - we should confirm it won't cause any problem (because we require it to return a successful status)

**Comment by ixarlie** [comment]
Hi @DavidGarciaCat , yeah I noticed that there are BC breaks too but I am not sure how to resolve it.

My project was using the version 2.x and we were having this same issue https://github.com/mailgun/mailgun-php/issues/577 . The version 2.x is using Http\Client\HttpClient instead of Psr\Http\Client\ClientInterface so everything was working well in regards to the http client.

This fix https://github.com/mailgun/mailgun-php/pull/590 is not resolved in 2.8.1, so we decided to upgrade to "3.2.x" (we do have some other dependencies conflicts right now so we cannot go further)

The version 3.2.x requires `"php-http/client-common": "^1.9 || ^2.0" `(my project is locked at `php-http/httplug v1.1.0`) but in the php-http/httplug 1.x the interface `Http\Client|HttpClient` is not extending `Psr\Http\Client\ClientInterface` as it does in 2.x.

I'd say this issue is not happening for mailgun-php versions > 3.2.x because from 3.3.0 composer is `php-http/client-common: ^2.2.1` which requires `php-http/httplug: ^2.0`

However, it is odd that the HttpApi is doing this logic and there is a BC in Mailgun\Api\Webhook when Webhook is extending HttpApi. Suppression is different because is not extending from HttpApi, so that's why I copied the same bits from HttpApi.
```php
    // https://github.com/mailgun/mailgun-php/blob/master/src/Api/HttpApi.php
    public function __construct($httpClient, RequestBuilder $requestBuilder, Hydrator $hydrator)
    {
        if (!is_a($httpClient, ClientInterface::class) &&
            !is_a($httpClient, PluginClient::class)) {
            throw new \RuntimeException('httpClient must be an instance of
            Psr\Http\Client\ClientInterface or Http\Client\Common\PluginClient');
        }
```
BTW, I think the code above is not necessary for versions greater than 3.3.0 and the type declaration can be restored to ClientInterface because from 3.3.0 as I commented before httplug is 2.x and PluginClient implements Http\Client|HttpClient which extends Psr\Http\Client\ClientInterface.


Probably we may not really need this PR and the linked issue can be closed, but then I really need https://github.com/mailgun/mailgun-php/pull/590 to be pushed in the 2.x branch and a new release 2.8.2 or 2.9.0 please.


Thanks in advanced!

**Comment by DavidGarciaCat** [comment]
Hey @Nyholm and @Pavlico 

What do you guys think might be better here? Moving that #590 to 2.x (as @ixarlie suggested) or keep moving forward with this PR?

Personally, I would like to reduce the number of versions we might need to maintain, but I would like to hear your point of view too.

Cheers,

**Comment by DavidGarciaCat** [comment]
And @ixarlie 

After checking the failing analysis here (https://github.com/mailgun/mailgun-php/pull/784/checks?check_run_id=3388496952) it seems to be quite clear the problem:

```
[BC] CHANGED: The parameter $httpClient of Mailgun\Api\Webhook#__construct() changed from Psr\Http\Client\ClientInterface to no type
[BC] CHANGED: The parameter $httpClient of Mailgun\Api\Suppression#__construct() changed from Psr\Http\Client\ClientInterface to no type
```

I believe (and I am not sure) it might be fixed if we run a conditional (such as `$client instanceof ...`) but we should test it.

Cheers,

**Comment by Nyholm** [comment]
The BC checking script is very strict. 
There is actually no scenario where a users code will break if we merge this. 

I’ll be happy to merge this pr and then tag a minor release.

**Comment by DavidGarciaCat** [comment]
> The BC checking script is very strict.
> There is actually no scenario where a users code will break if we merge this.
> 
> I’ll be happy to merge this pr and then tag a minor release.

Works for me. But my merge permissions don't allow me to merge skipping that check... Maybe you can help here?

File: src/Api/Suppression.php [hunk 2]
```diff
@@ -41,8 +42,13 @@ class Suppression
      */
     private $hydrator;
 
-    public function __construct(ClientInterface $httpClient, RequestBuilder $requestBuilder, Hydrator $hydrator)
+    public function __construct($httpClient, RequestBuilder $requestBuilder, Hydrator $hydrator)
     {
+        if (!is_a($httpClient, ClientInterface::class) &&
```
Comment (Nyholm) [inline src/Api/Suppression.php:47]: Let's change this to `instanceof` instead.

File: src/Api/Suppression.php [hunk 3]
```diff
@@ -41,8 +42,13 @@ class Suppression
      */
     private $hydrator;
 
-    public function __construct(ClientInterface $httpClient, RequestBuilder $requestBuilder, Hydrator $hydrator)
+    public function __construct($httpClient, RequestBuilder $requestBuilder, Hydrator $hydrator)
```
Comment (Nyholm) [inline src/Api/Suppression.php:45]: Lets add a doc block that describes `$httpClient`. Same on `Webhook`

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you. Only two small changes.

File: src/Api/Suppression.php (hunk 2 above)
Comment (ixarlie) [reply src/Api/Suppression.php:47]: Hi @Nyholm , should I do the same for `HttpApi` ?

https://github.com/mailgun/mailgun-php/blob/master/src/Api/HttpApi.php#L47-L53

Comment (Nyholm) [reply src/Api/Suppression.php:47]: Yes please

File: src/Api/Suppression.php (hunk 3 above)
Comment (ixarlie) [reply src/Api/Suppression.php:50]: The doc block was added

**Comment by ixarlie** [comment]
More failing checks than before :confused:

**Comment by Nyholm** [comment]
Thank you. 
I’m fine with the pr in its current state. I’ll merge it next time I’m at a computer. I’ll deal with updating the baseline and the failing checks.

**Comment by Pavlico** [comment]
Hi @Nyholm, should I try to add fixes to code so it passes the thingy or as you said you will merge it in this state :)

**Comment by DavidGarciaCat** [comment]
> Hi @Nyholm, should I try to add fixes to code so it passes the thingy or as you said you will merge it in this state :)

Hey @Pavlico 

If you have the time to work on it, then yes please!

Thank you!

**Comment by Pavlico** [comment]
Moved to #798

**Pull request closed without merging**